src/html-imports.js

```diff
@@ -9,718 +9,754 @@
-    if (text && ABS_URL_TEST.test(text)) {
-      return text;
-    } else {
-      return new URL(text, linkUrl).href;
-    }
   }
 
-};
-
-/********************* xhr processor *********************/
-var xhr = {
-  async: true,
-
-  ok: function(request) {
-    return (request.status >= 200 && request.status < 300)
-        || (request.status === 304)
-        || (request.status === 0);
-  },
-
-  load: function(url, next, nextContext) {
-    var request = new XMLHttpRequest();
-    if (flags.bust) {
-      url += '?' + Math.random();
-    }
-    request.open('GET', url, xhr.async);
-    request.addEventListener('readystatechange', function(e) {
-      if (request.readyState === 4) {
-        // Servers redirecting an import can add a Location header to help us
-        // polyfill correctly.
-        var redirectedUrl = null;
+  /********************* path fixup *********************/
+  const ABS_URL_TEST = /(^\/)|(^#)|(^[\w-\d]*:)/;
+  const CSS_URL_REGEXP = /(url\()([^)]*)(\))/g;
+  const CSS_IMPORT_REGEXP = /(@import[\s]+(?!url\())([^;]*)(;)/g;
+
+
+  // path fixup: style elements in imports must be made relative to the main
+  // document. We fixup url's in url() and @import.
+  const Path = {
+    resolveUrlsInStyle: function(style, linkUrl) {
+      style.textContent = Path.resolveUrlsInCssText(style.textContent, linkUrl);
+    },
+
+    resolveUrlsInCssText: function(cssText, linkUrl) {
+      let r = Path.replaceUrls(cssText, linkUrl, CSS_URL_REGEXP);
+      r = Path.replaceUrls(r, linkUrl, CSS_IMPORT_REGEXP);
+      return r;
+    },
+
+    replaceUrls: function(text, linkUrl, regexp) {
+      return text.replace(regexp, function(m, pre, url, post) {
+        let urlPath = url.replace(/["']/g, '');
+        if (linkUrl) {
+          urlPath = Path._resolveUrl(urlPath, linkUrl);
+        }
+        return pre + '\'' + urlPath + '\'' + post;
+      });
+    },
+
+    replaceAttrUrl: function(text, linkUrl) {
+      if (text && ABS_URL_TEST.test(text)) {
+        return text;
+      } else {
+        return Path._resolveUrl(text, linkUrl);
+      }
+    },
+
+    _resolveUrl: function(url, base) {
+      // Lazy feature detection.
+      if (Path.__workingURL === undefined) {
+        Path.__workingURL = false;
         try {
-          var locationHeader = request.getResponseHeader("Location");
-          if (locationHeader) {
-            redirectedUrl = (locationHeader.substr( 0, 1 ) === "/")
-              ? location.origin + locationHeader  // Location is a relative path
-              : locationHeader;                   // Full path
+          const u = new URL('b', 'http://a');
+          u.pathname = 'c%20d';
+          Path.__workingURL = (u.href === 'http://a/c%20d');
+        } catch (e) {}
+      }
+
+      if (Path.__workingURL) {
+        return (new URL(url, base)).href;
+      }
+
+      // Fallback to creating an anchor into a disconnected document.
+      let doc = Path.__tempDoc;
```

Comment:
Did some tests to compare creating 1000 URLs vs creating 1 doc and reusing it for urls, and url method (even when polyfilled) is ~1.5-2x more efficient. Tested with https://gist.github.com/valdrinkoshi/195c1dfc44ffc164b96c0d6cc1d65530